### PR TITLE
New module: manage 1&1 cloud compute  (cloud/oneandone/oneandone_server)

### PR DESCRIPTION
**Pull request by edevenport**
##### SUMMARY
This module adds support for the 1&1 Cloud Server platform. The oneandone submodule allows the virtual servers to be built and destroyed.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
oneandone

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION
Below is the output from test playbooks utilizing the new module:

```
Server create
=============

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Create a server] *********************************************************
changed: [localhost]

TASK [Print create server output] **********************************************
ok: [localhost] => {
    "msg": "Create server output: {u'changed': True, u'machines': [{u'cloudpanel_id': u'F59F06F', u'datacenter': {u'id': u'908DC2072407C94C8054610AD5A53B8C', u'country_code': u'US', u'location': u'United States of America'}, u'description': u'testing server creation with ansible', u'public_ipv4': u'74.208.131.201', u'private_networks': None, u'first_password': u'********', u'dvd': None, u'image': {u'id': u'8E3BAA98E3DFD37857810E0288DD8FBA', u'name': u'centos6-32min'}, u'alerts': [], u'creation_date': u'2017-05-23T17:35:30+00:00', u'hardware': {u'vcore': 2, u'ram': 1, u'cores_per_processor': 1, u'fixed_instance_size_id': None, u'hdds': [{u'is_main': True, u'id': u'C3B658910012ABC84BA021D677579914', u'size': 20}, {u'is_main': False, u'id': u'2F0F2119F6C9833821CC1B9484F3C657', u'size': 20}]}, u'ips': [{u'firewall_policy': {u'id': u'A4D6302FA44955C797772E34880CBA42', u'name': u'centos7-64cpanel'}, u'ip': u'74.208.131.201', u'reverse_dns': None, u'load_balancers': [], u'main': True, u'type': u'IPV4', u'id': u'955F9D1AF64BB502BCE6A6829AFAE9DD'}], u'status': {u'state': u'POWERED_ON', u'percent': None}, u'rsa_key': 0, u'snapshot': None, u'monitoring_policy': None, u'id': u'702BADAD7EACD7D00705F2589163F821', u'name': u'server01'}]}"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0   
```

```
Server stop
===========

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Stop the server] *********************************************************
changed: [localhost]

TASK [Print server stop output] ************************************************
ok: [localhost] => {
    "msg": "Server stop: {u'changed': True, u'machines': [{u'cloudpanel_id': u'F59F06F', u'datacenter': {u'id': u'908DC2072407C94C8054610AD5A53B8C', u'country_code': u'US', u'location': u'United States of America'}, u'description': u'testing server creation with ansible', u'public_ipv4': u'74.208.131.201', u'private_networks': None, u'first_password': u'********', u'dvd': None, u'image': {u'id': u'8E3BAA98E3DFD37857810E0288DD8FBA', u'name': u'centos6-32min'}, u'alerts': [], u'creation_date': u'2017-05-23T17:35:30+00:00', u'hardware': {u'vcore': 2, u'ram': 1, u'cores_per_processor': 1, u'fixed_instance_size_id': None, u'hdds': [{u'is_main': True, u'id': u'C3B658910012ABC84BA021D677579914', u'size': 20}, {u'is_main': False, u'id': u'2F0F2119F6C9833821CC1B9484F3C657', u'size': 20}]}, u'ips': [{u'firewall_policy': {u'id': u'A4D6302FA44955C797772E34880CBA42', u'name': u'centos7-64cpanel'}, u'ip': u'74.208.131.201', u'reverse_dns': None, u'load_balancers': [], u'main': True, u'type': u'IPV4', u'id': u'955F9D1AF64BB502BCE6A6829AFAE9DD'}], u'status': {u'state': u'POWERED_OFF', u'percent': None}, u'rsa_key': 0, u'snapshot': None, u'monitoring_policy': None, u'id': u'702BADAD7EACD7D00705F2589163F821', u'name': u'server01'}]}"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0   
```

```
Server start
============

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Start the server] ********************************************************
changed: [localhost]

TASK [Print server start output] ***********************************************
ok: [localhost] => {
    "msg": "Server start: {u'changed': True, u'machines': [{u'cloudpanel_id': u'F59F06F', u'datacenter': {u'id': u'908DC2072407C94C8054610AD5A53B8C', u'country_code': u'US', u'location': u'United States of America'}, u'description': u'testing server creation with ansible', u'public_ipv4': u'74.208.131.201', u'private_networks': None, u'first_password': u'********', u'dvd': None, u'image': {u'id': u'8E3BAA98E3DFD37857810E0288DD8FBA', u'name': u'centos6-32min'}, u'alerts': [], u'creation_date': u'2017-05-23T17:35:30+00:00', u'hardware': {u'vcore': 2, u'ram': 1, u'cores_per_processor': 1, u'fixed_instance_size_id': None, u'hdds': [{u'is_main': True, u'id': u'C3B658910012ABC84BA021D677579914', u'size': 20}, {u'is_main': False, u'id': u'2F0F2119F6C9833821CC1B9484F3C657', u'size': 20}]}, u'ips': [{u'firewall_policy': {u'id': u'A4D6302FA44955C797772E34880CBA42', u'name': u'centos7-64cpanel'}, u'ip': u'74.208.131.201', u'reverse_dns': None, u'load_balancers': [], u'main': True, u'type': u'IPV4', u'id': u'955F9D1AF64BB502BCE6A6829AFAE9DD'}], u'status': {u'state': u'POWERED_ON', u'percent': 53}, u'rsa_key': 0, u'snapshot': None, u'monitoring_policy': None, u'id': u'702BADAD7EACD7D00705F2589163F821', u'name': u'server01'}]}"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0  
```

```
Server delete
=============

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Delete a server] *********************************************************
changed: [localhost]

TASK [Print server delete output] **********************************************
ok: [localhost] => {
    "msg": "Server delete: {u'changed': True, u'machines': [{u'hostname': u'server01', u'id': u'702BADAD7EACD7D00705F2589163F821'}]}"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0 
```

### CODE REVIEW

**Comment by resmo** [comment]
Thanks for your effort! I didn't yet look much at the code but already have a few general hints:

## Naming considerations
Since it looks like 1&1 has some more APIs (https://cloudpanel-api.1and1.com/documentation/v1/en/examples.html), I would rename the module to reflect this modules only handles the server API. e.g. ono_server. This let you add more modules handling different APIs like firewall: ono_firewall_policy, ono_firewall_rule, ono_lb, ono_lb_rule, etc.

## Common code in utils
If you have shared code e.g. authentication you want to use for a few modules related to 1&1, put this code into https://github.com/ansible/ansible/tree/devel/lib/ansible/module_utils and include it in your modules. See others have done it. 

## Integration Tests
Test your modules by adding integration tests, especially idempotency. Because the 1&1 is proprietary, I would help to share results along with the modules PR (as comments)

File: lib/ansible/modules/cloud/oneandone/oneandone.py [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+DOCUMENTATION = '''
+---
+module: oneandone
+short_description: create, destroy, start, stop, and reboot a 1&1 Host machine.
+description:
+     - create, destroy, update, start, stop, and reboot a 1&1 Host machine.
+       When the machine is created it can optionally wait for it to be
+       'running' before returning.
+       This module has a dependency on 1and1 >= 1.0
+version_added: "2.1"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/oneandone/oneandone.py]: `version_added: "2.4"`

File: lib/ansible/modules/cloud/oneandone/oneandone.py [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+
+DOCUMENTATION = '''
+---
+module: oneandone
+short_description: create, destroy, start, stop, and reboot a 1&1 Host machine.
+description:
+     - create, destroy, update, start, stop, and reboot a 1&1 Host machine.
+       When the machine is created it can optionally wait for it to be
+       'running' before returning.
+       This module has a dependency on 1and1 >= 1.0
```
Comment (gundalow) [inline lib/ansible/modules/cloud/oneandone/oneandone.py]: You can remove this line as it's in the `requirements:` block

File: lib/ansible/modules/cloud/oneandone/oneandone.py [hunk 3]
```diff
@@ -0,0 +1,793 @@
...
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+module: oneandone
+short_description: create, destroy, start, stop, and reboot a 1&1 Host machine.
+description:
+     - create, destroy, update, start, stop, and reboot a 1&1 Host machine.
```
Comment (gundalow) [inline lib/ansible/modules/cloud/oneandone/oneandone.py]: `Create...`

File: lib/ansible/modules/cloud/oneandone/oneandone.py [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
+      - List of machine IDs or hostnames.
+    required: false for 'running' state, true otherwise
+  count:
+    description:
+      - The number of machines to create.
+    required: false
+    default: 1
+  ssh_key:
+    description:
+      - User's public SSH key.
```
Comment (gundalow) [inline lib/ansible/modules/cloud/oneandone/oneandone.py]: Is this a path to a key, or the contents of the key. Please update `description:` and example to be clearer.

File: lib/ansible/modules/cloud/oneandone/oneandone.py [hunk 5]
```diff
@@ -0,0 +1,793 @@
...
+            (changed, machines) = create_machine(module, oneandone_conn)
+        except Exception as e:
+            module.fail_json(msg=str(e))
+
+    module.exit_json(changed=changed, machines=machines)
+
+
+from ansible.module_utils.basic import AnsibleModule
+
+main()
```
Comment (gundalow) [inline lib/ansible/modules/cloud/oneandone/oneandone.py]: Please write as
```
if __name__ == '__main__':
    main()
```

File: lib/ansible/modules/cloud/oneandone/oneandone.py [hunk 6]
```diff
@@ -0,0 +1,793 @@
...
+
+    elif state in ('running', 'stopped'):
+        try:
+            (changed, machines) = startstop_machine(module, oneandone_conn)
+        except Exception as e:
+            module.fail_json(msg=str(e))
+
+    elif state == 'present':
+        if not _validate_hardware_params(module):
+            module.fail_json(
```
Comment (gundalow) [inline lib/ansible/modules/cloud/oneandone/oneandone.py]: There are built in helper functions for option validation, see the following draft documentation:
https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/oneandone/oneandone.py (hunk 5 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone.py]: Thank you for the feedback. The requested changes have been made.

**Comment by edevenport** [comment]
@resmo @gundalow Thank you both for you feedback and review. The review items have been addressed and we renamed the module as suggested. We are currently in the process of consolidating some common functions under the module_utils and should have that completed shortly.

@resmo With regards to integration tests, are you suggesting the results of a successful playbook run using the module? Or are you suggesting something more specific?

Once this particular server submodule is approved, we already have other 1&1 API operations (including private_network, firewall_policy, loadbalancers, etc.) prepared as additional submodules under other branches. We should have nearly complete coverage.

**Comment by edevenport** [comment]
@resmo @gundalow We have consolidated common functions in a module_utils file and I attached the output from our test playbooks in the PR description. That should hopefully address all the outstanding feedback.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+      - When creating multiple machines at once, whether to differentiate
+        hostnames by appending a count after them or substituting the count
+        where there is a %02d or %03d in the hostname string.
+    default: "yes"
+    choices: [ "yes", "no" ]
+
+requirements:
+  - "1and1"
+  - "python >= 2.6"
+author: "Amel Ajdinovic (@aajdinov), Ethan Devenport (@edevenport)"
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: please create a list: e.g. 

~~~ yaml
author: 
  -  "Amel Ajdinovic (@aajdinov)" 
  -  "Ethan Devenport (@edevenport)" 
~~~

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+    auth_token: oneandone_private_api_key
+    state: stopped
+    instance_ids:
+      - 'node01'
+      - 'node02'
+      - 'node03'
+'''
+
+RETURN = '''
+changed:
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: this is a common return and must not be documented in a module. Please remove it.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+    required: true
+  description:
+    description:
+      - The description of the machine.
+    required: false
+  appliance:
+    description:
+      - The operating system name or ID for the machine.
+        It is required only for 'present' state.
+    required: true
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: the requirement of this identical to the modules arguments requirements (required=True), if a param if not always required, set `required: false`. The description has all information.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+  datacenter:
+    description:
+      - The datacenter location.
+    required: false
+    default: US
+    choices: [ "US", "ES", "DE", "GB" ]
+  hostname:
+    description:
+      - The hostname or ID of the machine. Only used when state is 'present'.
+    required: true
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: set `required: false` please.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+      - The instance size name or ID of the machine.
+        It is required only for 'present' state, and it is mutually exclusive with
+        vcore, cores_per_processor, ram, and hdds parameters.
+    required: true
+    choices: [ "S", "M", "L", "XL", "XXL", "3XL", "4XL", "5XL" ]
+  vcore:
+    description:
+      - The total number of processors.
+        It must be provided with cores_per_processor, ram, and hdds parameters.
+    required: true
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: required: false, only required on state=present?

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 12]
```diff
@@ -0,0 +1,657 @@
...
+  cores_per_processor:
+    description:
+      - The number of cores per processor.
+        It must be provided with vcore, ram, and hdds parameters.
+    required: true
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: required: false, only required on state=present?

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 13]
```diff
@@ -0,0 +1,657 @@
...
+  ram:
+    description:
+      - The amount of RAM memory.
+        It must be provided with with vcore, cores_per_processor, and hdds parameters.
+    required: true
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: required: false, only required on state=present?

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 14]
```diff
@@ -0,0 +1,657 @@
...
+  hdds:
+    description:
+      - A list of hard disks with nested "size" and "is_main" properties.
+        It must be provided with vcore, cores_per_processor, and ram parameters.
+    required: true
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: required: false, only required on state=present?

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 15]
```diff
@@ -0,0 +1,657 @@
...
+      - The load balancer name or ID.
+    required: false
+  monitoring_policy:
+    description:
+      - The monitoring policy name or ID.
+    required: false
+  instance_ids:
+    description:
+      - List of machine IDs or hostnames. It is required for all states except 'running'.
+    required: true
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: set required: false

**Review by resmo — CHANGES_REQUESTED** [review]
A few notes about the module docs. 

The modules handles multiple resources at once, while there is no policy against that, I just wanted to point out this tends to makes things more complicated, e.g. which node changed? returns does return a list of each node? Ansible core has already support for loops built-in (with_ loops). I would reconsider to simply the module in a way to process one resource at once, unless there is a good reason against it.

Further, it seems to me the module does not or not always handle idempotency, e.g. state=present, does it? if you add an identical task in you test playbook after the create task, would it be changed=False? Not handling idempotency is a no go for me (unless there is a good reason)

It also seems to me to implement check mode support would be possible.

**Review by gundalow — APPROVED** [review]

**Comment by edevenport** [comment]
Hello @resmo - @aajdinov has made the requested changes.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 10 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: Thank you for the review. I've set `required: false`.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 9 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: I've updated `required` to `false`.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 11 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: Thanks for the tip. It is now set to `required: false`.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 12 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: I've changed it to `required: false`.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 13 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: `required` is now `false`. Thanks.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 14 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: I've updated `required` to `false`.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 15 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: I've set `required: false`. Thank you.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 7 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py]: Thank you for the review. I've updated authors to a list format.

**Comment by aajdinov** [comment]
Hello @resmo . Thank you for the review. The proposed changes ware made and pushed to repository.

**Comment by edevenport** [comment]
@resmo Can we get a sign off on the changes?

**Comment by edevenport** [comment]
@resmo Any update?

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 16]
```diff
@@ -0,0 +1,644 @@
...
+            datacenter=dict(
+                choices=DATACENTERS,
+                default='US'),
+            private_network=dict(type='str'),
+            firewall_policy=dict(type='str'),
+            load_balancer=dict(type='str'),
+            monitoring_policy=dict(type='str'),
+            wait=dict(type='bool', default=True),
+            wait_timeout=dict(type='int', default=600),
+            state=dict(type='str', default='present'),
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: please add a list of choices here: 

~~~diff
-            state=dict(type='str', default='present'),
+            state=dict(type='str', default='present', choices=['present', 'absent', 'running', 'stopped']),
~~~

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 17]
```diff
@@ -0,0 +1,644 @@
...
+            hostname=dict(type='str'),
+            description=dict(type='str'),
+            appliance=dict(type='str'),
+            fixed_instance_size=dict(type='str'),
+            vcore=dict(type='int'),
+            cores_per_processor=dict(type='int'),
+            ram=dict(type='float'),
+            hdds=dict(type='list'),
+            count=dict(type='int', default=1),
+            ssh_key=dict(type='raw', default=None),
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py]: `default=None` is the default, so doesn't have to be set here

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py [hunk 18]
```diff
@@ -0,0 +1,644 @@
...
+                oneandone_conn.modify_server_status(
+                    server_id=server['id'],
+                    action='POWER_OFF',
+                    method='SOFTWARE')
+            elif state == 'running' and server['status']['state'] == 'POWERED_OFF':
+                oneandone_conn.modify_server_status(
+                    server_id=server['id'],
+                    action='POWER_ON',
+                    method='SOFTWARE')
+        except Exception as ex:
```
Comment (resmo) [inline lib/ansible/modules/cloud/oneandone/oneandone_server.py:536]: Catch all exceptions should be avoided but it is not a "blocker"

**Review by resmo — APPROVED** [review]
I only have a few tweaks noted. See comments inline.

The module is a good start but I also see some potential e.g. consistent returns: "return back the resource (e.g. server dict) to the user, check mode support.

File: lib/ansible/modules/cloud/oneandone/oneandone_server.py (hunk 18 above)
Comment (aajdinov) [reply lib/ansible/modules/cloud/oneandone/oneandone_server.py:536]: Hello @resmo 
Thank you for your feedback. Unfortunately, the 1&1 python SDK is currently throwing a generic `Exception` in `modify_server_status` method, which is why we catch it here.

**Comment by aajdinov** [comment]
Hello @resmo .

I have added the check mode to the module and addressed the other suggestions you made.
Now I am a bit puzzled by the checks failing (all of them on items that were not altered in my latest commit).
`enum` and `xrange` are used and are working properly in the module.
Have there been any changes to the way shippable checks the new modules that could have caused this PR to fail checks?

**Comment by mattclay** [comment]
@edevenport We recently added several new tests which this PR is now failing on.

You can find documentation (currently a work in progress) for the sanity tests here: http://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/

**Comment by alibazlamit** [comment]
@mattclay Thanks for informing us about the new tests. 
Now that we have addressed all issues reported by the sanity tests the build is still failing but without indicating which tests are failing or what is wrong, could you help us find what is causing the build to fail?

This is what i could find in the console log but i am not sure what is going on
```
2017-09-14 11:14:26 Run command: importer.py lib/ansible/module_utils/oneandone.py lib/ansible/modules/cloud/oneandone/__init__.py lib/ansible/modules/cloud/oneandone/oneandone_server.py
2017-09-14 11:14:26 Traceback (most recent call last):
2017-09-14 11:14:26   File "/root/src/github.com/ansible/ansible/test/runner/test.py", line 669, in <module>
2017-09-14 11:14:26     main()
2017-09-14 11:14:26   File "/root/src/github.com/ansible/ansible/test/runner/test.py", line 89, in main
2017-09-14 11:14:26     args.func(config)
2017-09-14 11:14:26   File "/root/src/github.com/ansible/ansible/test/runner/lib/sanity/__init__.py", line 99, in command_sanity
2017-09-14 11:14:26     result = test.test(args, targets, python_version=version)
2017-09-14 11:14:26   File "/root/src/github.com/ansible/ansible/test/runner/lib/sanity/import.py", line 104, in test
2017-09-14 11:14:26     results = [re.search(pattern, line).groupdict() for line in ex.stdout.splitlines()]
2017-09-14 11:14:26 AttributeError: 'NoneType' object has no attribute 'groupdict'
2017-09-14 11:14:26 ERROR: InvocationError: '/root/src/github.com/ansible/ansible/test/runner/test.py sanity --failure-ok -v --junit --test ansible-doc --test import --coverage --require lib/ansible/module_utils/oneandone.py --require lib/ansible/modules/cloud/oneandone/__init__.py --require lib/ansible/modules/cloud/oneandone/oneandone_server.py --metadata metadata-AD02P7.json --color yes --requirements --base-branch origin/devel --python 2.6 --coverage-label tox-2.6'
2017-09-14 11:14:26 ___________________________________ summary ____________________________________
2017-09-14 11:14:26 ERROR:   py26: commands failed
2017-09-14 11:14:26 ERROR: Command "tox -c test/runner/tox.ini -e py26 -- /root/src/github.com/ansible/ansible/test/runner/test.py sanity --failure-ok -v --junit --test ansible-doc --test import --coverage --require lib/ansible/module_utils/oneandone.py --require lib/ansible/modules/cloud/oneandone/__init__.py --require lib/ansible/modules/cloud/oneandone/oneandone_server.py --metadata metadata-AD02P7.json --color yes --requirements --base-branch origin/devel --python 2.6 --coverage-label tox-2.6" returned exit status 1.

```

**Comment by mattclay** [comment]
@alibazlamit That's a bug in our test infrastructure. I'll work on a fix.

**Comment by mattclay** [comment]
@alibazlamit I've fixed the issue in our test infrastructure and restarted the tests.

**Comment by alibazlamit** [comment]
@mattclay all issues are resolved can we move ahead with this PR?

Thanks.

**Comment by aajdinov** [comment]
@resmo @gundalow @mattclay 
Can we get a sign off on the changes? Thank you.

**Review by gundalow — APPROVED** [review]

**Comment by aajdinov** [comment]
@gundalow 
Thank you for approving the changes. Is there anything else that needs to happen for this PR to be accepted/merged into the main branch?

**Review by resmo — APPROVED** [review]
shipit

**Comment by gundalow** [comment]
Merged into `devel` will be released in `2.5.0`
Thanks to all that have worked on this.

**Pull request merged**